### PR TITLE
[Generative Score API] Optimization to Remove Decode

### PR DESCRIPTION
**Pull request by sundar24295s**
## 🚀 Motivation

- This PR is a follow-up to the Decoder-only Scoring API introduced in [PR #6460](https://github.com/sgl-project/sglang/pull/6460), which was initially proposed and discussed in [Issue #5973](https://github.com/sgl-project/sglang/issues/5973).
- Achieved a **40% reduction in single-request latency** for a 300-token input (120 query tokens + 180 item tokens), decreasing from 33 ms to 20 ms.

## 🔧 Modifications

- Introduces dedicated objects for the Scoring API to simplify the implementation and enable targeted optimizations.

- **Remote Decode Optimization**:  
  Scoring workloads typically require only log probabilities at specific token positions and do not involve token generation. This PR removes the decode phase entirely for such workloads, resulting in significantly lower latency and improved throughput.

- The new code structure cleanly separates scoring-specific logic from the text generation path, ensuring that prefill-only optimizations do not interfere with generation behavior.

## 📈 Future Optimizations

This refactor lays the foundation for further improvements, including:

- **Memory Transfer Optimization**:  
  Move all CPU-GPU synchronization to the post-processing loop to better overlap `run_batch` with post-processing.

- **Multi-Item Scoring**:  
  Support scoring multiple items within a single prompt using custom attention masks via FlashInfer.  
  → See [flashinfer-ai/flashinfer#1015](https://github.com/flashinfer-ai/flashinfer/pull/1015)

## Breaking Changes
- None for public APIs - the `/v1/score` endpoint maintains backward compatibility
- Internal request handling has been refactored but doesn't affect external interfaces

## Accuracy Test
- TBA

##  Profiling
- **Updated Benchmark Scripts**: Enhanced `bench_score.py` to properly test the new scoring pipeline
- Before this change, for a single request we see `forward_extend` and `forward_decode`.
image
- After this change, we only have `forward_extend`.
image

## 🧪 Benchmark Comparison: Qwen3-0.5B on H100 (CUDA 12.8)

**Setup**:  
- Model: Qwen3-0.5B (pruned, decoder-only)  
- Prompt length: 300 tokens  
- Hardware: H100 GPU  
- Duration: 120s  
- Target RPS: 70  
- Item Count: 10  
- Distribution: Poisson  
- Transport: HTTP  

## Results

- Achieved a **40% reduction in single-request latency** for a 300-token input (120 query tokens + 180 item tokens), decreasing from 33 ms to 20 ms.
- For batch size 10 and the same 300-token input, the table below compares the latency percentiles before and after the change.

### 🔍 Summary of Improvement

| Metric                  | Baseline | With Change | Improvement     |
|-------------------------|----------|-------------|-----------------|
| Avg Response Time (ms)  | 83.92    | 51.16       | ↓ 39%           |
| P50 Latency (ms)        | 51.77    | 39.10       | ↓ 24%           |
| P90 Latency (ms)        | 114.13   | 73.37       | ↓ 36%           |
| P99 Latency (ms)        | 621.66   | 391.96      | ↓ 37%           |
| Single Request Latency  | 33 ms    | 20 ms       | ↓ 39%           |

---

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

File: python/sglang/srt/managers/scheduler_output_processor_mixin.py [hunk 1]
```diff
@@ -295,6 +300,120 @@ def process_batch_result_decode(
...
+                if logits_output.input_token_logprobs is not None:
+                    logits_output.input_token_logprobs = tuple(
+                        logits_output.input_token_logprobs.tolist()
+                    )
+        
+        # Free tokens from the memory pool (same as decode processing)
+        self.token_to_kv_pool_allocator.free_group_begin()
+        
+        # Process logprobs for scoring requests
+        if batch.return_logprob and logits_output is not None:
```
Comment (lanking520) [inline python/sglang/srt/managers/scheduler_output_processor_mixin.py:332]:
```suggestion
        if batch.return_logprob and logits_output:
```

File: python/sglang/srt/managers/scheduler_output_processor_mixin.py [hunk 2]
```diff
@@ -295,6 +300,120 @@ def process_batch_result_decode(
...
+        without generating new tokens. They are marked as finished in run_batch.
+        """
+        logits_output = result.logits_output
+        
+        if self.enable_overlap:
+            # Resolve the last batch result for overlap scheduling
+            logits_output, _, _ = self.tp_worker.resolve_last_batch_result(launch_done)
+        else:
+            # Move logprobs to CPU if needed
+            if batch.return_logprob and logits_output is not None:
```
Comment (lanking520) [inline python/sglang/srt/managers/scheduler_output_processor_mixin.py:322]:
```suggestion
            if batch.return_logprob and logits_output:
```

File: python/sglang/srt/managers/scheduler_output_processor_mixin.py [hunk 3]
```diff
@@ -295,6 +300,120 @@ def process_batch_result_decode(
...
+        if batch.return_logprob and logits_output is not None:
+            for i, req in enumerate(batch.reqs):
+                if req.is_scoring_request:
+                    # For scoring requests, we use next_token logprobs since we want 
+                    # to know the probability of specific tokens at the next position
+                    if (logits_output.next_token_token_ids_logprobs_val is not None and
+                        logits_output.next_token_token_ids_logprobs_idx is not None):
+                        
+                        # Initialize all the logprob fields for scoring request
+                        if req.input_token_logprobs_val is None:
```
Comment (lanking520) [inline python/sglang/srt/managers/scheduler_output_processor_mixin.py:341]: what is the consequence if we just pass in None value as input? Should we just change the default value of request to [] if needed?
